### DISCUSSION TITLE
Request parameters model not exported to docs (and OpenAPI Specification)

### DISCUSSION THREAD

**Discussion by Philipp3211** [fastapi/fastapi#11543] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI, Depends
from pydantic import BaseModel


class ParamsModel(BaseModel):
    foo: str
    bar: str


app = FastAPI()


@app.get("/")
def test_route(params: Annotated[ParamsModel, Depends()]):
    return params.foo + params.bar
```

### Description

Start the app then open the browser and go to `/docs` and `/openapi.json` respectively

What is currently:
* `/docs`: `ParamsModel` is not present in the Schemas. Right now only `HTTPValidationError` and `ValidationError` are in the schemas.
* `/openapi.json`: `ParamsModel` is not present in the OpenAPI-Spec either

What I'm expecting:
* I want the `ParamsModel` to be present in the schemas and the OpenAPI-Spec

_Why do I need this?_
The type is required for my UI being able to handle the request data by using OpenAPI exported types only.

I need foo and bar as parameters because on UI side I use caching that depends on foo and bar being present in the URL.

Is there a way to get this behavior with FastAPI?

### Operating System

Linux, Windows

### FastAPI Version

0.110.2

### Pydantic Version

1.10.2

### Python Version

Python 3.9.5

**Comment by PhysicallyActive** [comment]
Why do you need the use of `Depends`? Shouldn't this solve your issue

```python
from fastapi import FastAPI
from pydantic import BaseModel


class ParamsModel(BaseModel):
    foo: str
    bar: str


app = FastAPI()


@app.post("/")
def test_route(params: ParamsModel):
    return params.foo + params.bar
```

**Reply by Philipp3211** [reply]
I need `foo` and `bar` as parameters because on UI side I use caching that depends on `foo` and `bar` being present in the URL. Using your solution makes them move from parameters into body.

**Reply by YuriiMotov** [reply]
Do you want `params` to be a Query parameters or Body parameters?
To make them body parameters just use `params: ParamsModel` annotation.
If you want to make them Query parameters - you can only do this using dataclasses. FastAPI doesn't have official support of Pydantic models for Query parameters

**Reply by Philipp3211** [reply]
I need them as query parameters. FastAPI seems to be supporting [classes for query parameters](https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/?h=#classes-as-dependencies_1). It's just that the classes are not exported as schemas to the docs.

**Reply by PhysicallyActive** [reply]
If you really need to export the schema with the docs, you can include your own specific openapi_schema:

```python
from typing import Annotated

from fastapi import Depends, FastAPI
from fastapi.openapi.utils import get_openapi
from pydantic import BaseModel


class ParamsModel(BaseModel):
    foo: str
    bar: str


app = FastAPI()


@app.post("/")
def test_route(params: Annotated[ParamsModel, Depends()]) -> str:
    return params.foo + params.bar

def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title="Your Title",
        version="0.1.0",
        routes=app.routes,
    )
    openapi_schema["components"]["schemas"][ParamsModel.__name__] = ParamsModel.model_json_schema()
    app.openapi_schema = openapi_schema
    return app.openapi_schema


app.openapi = custom_openapi
```

All according to this page: https://fastapi.tiangolo.com/how-to/extending-openapi/

**Reply by YuriiMotov** [reply]
> I need them as query parameters. FastAPI seems to be supporting [classes for query parameters](https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/?h=#classes-as-dependencies_1). It's just that the classes are not exported as schemas to the docs.

Yes, you can use either usual classes or dataclasses, but not Pydantic models for Query parameters.
Take a look at the answer of Sebastián Ramírez to this comment: https://github.com/tiangolo/fastapi/discussions/9709#discussioncomment-6532693, also https://github.com/tiangolo/fastapi/discussions/9709#discussioncomment-6624315

**Reply by Philipp3211** [reply]
You are right, pydantic models are not supported. However even after switching to usual class or dataclass, the model still isn't exported to the schemas or openapi specification.

**Reply by YuriiMotov** [reply]
```
from dataclasses import dataclass
from typing import Annotated

from fastapi import Depends, FastAPI


@dataclass
class ParamsModel:
    foo: str
    bar: str


app = FastAPI()


@app.get("/")
def test_route(params: Annotated[ParamsModel, Depends()]):
    return params.foo + params.bar
```

In the Swagger docs it looks as follows:
![dataclass](https://github.com/tiangolo/fastapi/assets/109919500/e90b6717-e08d-4f13-8bd5-644d77db434d)

Is it what you need? Or do you expect something else?

**Reply by Philipp3211** [reply]
To quote from my question above:
> What is currently:
> /docs: ParamsModel is not present in the Schemas. Right now only HTTPValidationError and ValidationError are in the schemas.
> /openapi.json: ParamsModel is not present in the OpenAPI-Spec either
> 
> What I'm expecting:
> I want the ParamsModel to be present in the schemas and the OpenAPI-Spec

**Reply by YuriiMotov** [reply, marked as answer]
Oh, now I see.. Sorry for the inattention.

I don't think you can easily do what you want in current version of FastAPI since it only accepts simple types (and arrays of simple type parameters) as a Query parameters.

When you use annotation like
```params: Annotated[ParamsModel, Depends()]```
it is just a shortcut to
```params: Annotated[ParamsModel, Depends(ParamsModel)]```

FastAPI just looks at the parameters of ParamsModel's __init__ method and treats them as subdependencies. So, it's equivalent of 
```
def get_query_params_dep(foo: str, bar: str) -> ParamsModel:
    return ParamsModel(foo=foo, bar=bar)

@app.get("/")
def test_route(params: Annotated[ParamsModel, Depends(get_query_params_dep)]):
    return params.foo + params.bar
```

I think you can (?) only achieve what you want by modifying openapi schema. Take a look at this:
```
from dataclasses import dataclass
from typing import Annotated

from fastapi import Depends, FastAPI
from fastapi.openapi.utils import get_openapi
from pydantic import TypeAdapter

app = FastAPI()

@dataclass
class ParamsModel:
    foo: str
    bar: str

@app.get("/")
def test_route(params: Annotated[ParamsModel, Depends()]):
    return params.foo + params.bar


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title=app.title,
        version=app.version,
        routes=app.routes
    )
    openapi_schema["components"]["schemas"]["ParamsModel"] = TypeAdapter(ParamsModel).json_schema()
    
    path_params = openapi_schema["paths"]["/"]["get"]["parameters"]
    EXCLUDE_PARAMS = ("foo", "bar")
    path_params = [param for param in path_params if param["name"] not in EXCLUDE_PARAMS]
    path_params.append(
        {
            "name": "params",
            "in": "query",
            "schema": {"type": "object", "ref": "#/components/schemas/ParamsModel"},
            "style": "form",
            "explode": True
        }
    )
    openapi_schema["paths"]["/"]["get"]["parameters"] = path_params
    
    app.openapi_schema = openapi_schema
    return app.openapi_schema

app.openapi = custom_openapi
```

This produces the following schema for path:
```
"/": {
    "get": {
        "summary": "Test Route",
        "operationId": "test_route__get",
        "parameters": [
            {
                "name": "params",
                "in": "query",
                "schema": {
                    "type": "object",
                    "ref": "#/components/schemas/ParamsModel"
                },
                "style": "form",
                "explode": true
            }
        ],
...
```

But pay attention, that FastAPI docs doesn't declare supporting dataclasses as Query parameters. It works for simple cases, but you may have troubles in more complex cases.

**Reply by Philipp3211** [reply]
> Yes, you can use either usual classes or dataclasses, but not Pydantic models for Query parameters.

Where did you read that dataclasses are allowed for query parameters? I can't find any info on this.

**Reply by YuriiMotov** [reply]
I didn't find anything in the documentation about supporting dataclasses as Query parameters, but at least in simple cases it works.
Dataclasses are useful here, because they create __init__ method automatically, so you don't need to write it manually.

**Reply by Philipp3211** [reply]
Well pydantic models also work for me, but they are explicitly not supported as stated in the link from you https://github.com/tiangolo/fastapi/discussions/9709#discussioncomment-6640931

I'm gonna stick with the documentation and use a casual class. If you want you can edit your answer and I will accept it as an answer.

**Reply by Philipp3211** [reply]
Forget what I just wrote. The solution doesn't work without the `@dataclass` decorator. If I remove it, I get `Unable to generate pydantic-core schema`.
